src/main/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessor.groovy

```diff
@@ -84,81 +76,80 @@ class JcrNodesProcessor implements ItemProcessor<JcrNode, Node> {
+            protoNodeBuilder.addMandatoryChildNode(buildNode(it, getMandatoryChildNodes(it)))
         }
-        return nodeBuilder.build()
+        return protoNodeBuilder.build()
     }
 
     /**
      * Identify all required child nodes
-     * @param JCR Node
-     * @return list of required child nodes
+     * @param jcrNode
+     * @return Collection of required child nodes or null if jcrNode has none
      */
-    private static Collection<JcrNode> getRequiredChildNodes(JcrNode jcrNode) {
-        return hasMandatoryChildNodes(jcrNode) ? jcrNode.getNodes().findAll{JcrNode childJcrNode -> isRequiredNode(childJcrNode)} : null
+    @Nullable
+    private static Collection<JcrNode> getMandatoryChildNodes(JcrNode jcrNode) {
+        return hasMandatoryChildNodes(jcrNode) ? jcrNode.getNodes().findAll{JcrNode childJcrNode -> isMandatoryNode(childJcrNode)} : null
     }
 
     /**
-     * Checks for mandatory child nodes
-     * @param JCR Node
-     * @return boolean flag for mandatory child node
+     * Determines if a JCR Node has mandatory child nodes
+     * @param jcrNode
      */
     private static boolean hasMandatoryChildNodes(JcrNode jcrNode) {
         return jcrNode.primaryNodeType.childNodeDefinitions.any { ((ItemDefinition)it).mandatory }
     }
 
     /**
-     * Checks required node
-     * @param JCR node
-     * @return boolean flag for mandatory node
+     * Checks to see if a JCR Node is a mandatory node
+     * @param node to check
      */
-    private static boolean isRequiredNode(JcrNode node) {
+    private static boolean isMandatoryNode(JcrNode node) {
         return node.definition.isMandatory()
     }
 
     /**
-     * Build Proto Property object from JCR node
-     * @param jcrNode for Proto properties
-     * @return Proto Property object
+     * Build a collection of Proto Properties from JCR node
+     * @param jcrNode to gather properties from
+     * @return resulting collection of transferable proto properties collected from jcrNode
      */
-    private static NodeProtos.Properties buildProperties(JcrNode jcrNode) {
+    @Nonnull
+    private static Collection<ProtoProperty> getProtoPropertiesFrom(JcrNode jcrNode) {
         final List<JcrProperty> properties = jcrNode.properties.toList()
-        Properties.Builder propertiesBuilder = Properties.newBuilder()
-        properties.each { JcrProperty jcrProperty ->
+        final transferableProperties = properties.findAll { JcrProperty jcrProperty ->
             //Before adding a property to the Current Node Proto message, check if the property
             //is Valid and if it should be actually sent to the client
-            if (isPropertyTransferable(jcrProperty)) {
-                Property property = toProperty(jcrProperty)
-                propertiesBuilder.addProperty(property)
-            }
+            isPropertyTransferable(jcrProperty)
+        }
+        return transferableProperties.collect { JcrProperty jcrProperty ->
```

Comment:
I prefer the functional style too.. The problem of null can be taken care by using findResults instead of collect.. It essentially does the same thing + filter null values...